semantic-tags/src/Tags/Tagging/Precise.hs

```diff
@@ -43,36 +64,133 @@ firstLine :: Source -> Text
 firstLine = Text.takeWhile (/= '\n') . toText . Source.take 180
 
 
--- FIXME: move GFoldable1 into semantic-ast.
-class GFoldable1 c t where
-  -- | Generically map functions over fields of kind @* -> *@, monoidally combining the results.
-  gfoldMap1
-    :: Monoid b
-    => (forall f . c f => f a -> b)
+-- FIXME: move Traversable1 into semantic-ast.
+-- FIXME: derive Traversable1 instances for TH-generated syntax types.
+
+-- | Simultaneous traversal of subterms of kind @*@ and @* -> *@ in an 'Applicative' context.
+--
+-- 'Traversable1' can express any combination of first- and second-order mapping, folding, and traversal.
+class Traversable1 c t where
```

Comment:
I’m really quite pleased with this class. It:

1. Subsumes the prior `GFoldable1` functionality.
2. Can express `Foldable` (when applied recursively).
3. Can express `Functor` (when applied recursively).
4. Can express `Traversable` (when applied recursively).
5. Suffices to express all of the tagging traversals.

The only bit I don’t really love is the name. There are two distinct uses of the suffix `1` in `base` that I’m aware of:

1. Classes such as `Generic1` & `Show1`, where it indicates that a similar class applied to kind `*` has been lifted to kind `* -> *`. This usage is most analogous to how we’re using it for `Traversable1`, and `base` does not define `Traversable1`, `Functor1`, or `Foldable1`.

2. Functions such as `foldl1` and `foldr1`, where it indicates that the function expects a non-empty collection. This use is further applied in the widely-depended-upon (directly or indirectly) [`semigroupoids` package](http://hackage.haskell.org/package/semigroupoids), which supplies `Foldable1` and `Traversable1` classes modelling non-empty foldable and traversable containers. `Data.List.NonEmpty` was moved into `base`, and it seems plausible that these might eventually be as well, which makes me a bit reluctant to use this name.

On the other hand, in `fused-effects` and `fused-syntax` we use the prefix `H` to denote higher-order analogues of the un-prefixed name, e.g. `HFunctor` is analogous to `Functor` but for `((* -> *) -> (* -> *)`. But the meaning here is different, too: the subterms are all of a homogeneous parameter type, not heterogeneous un-parameterized (constant) types. Our `Traversable1` cannot change the types of subterms, only of their annotations, which means that it’s a strictly different usage, somewhere between `H*` and the [`mono-traversable` package](http://hackage.haskell.org/package/mono-traversable).

If we do call it `Traversable1`, and publish it, and so forth, I suppose the worst that can happen is that we might occasionally have to qualify which one’s name we mean—I don’t count it as terribly likely that `Foldable1` or `Traversable1` will ever end up in `base`. So maybe this name is ok, and I’m just overthinking it?

@github/semantic-code: I would appreciate your thoughts.

Reply:
I think `Traversable1` is fine if we keep it in these functions. Maybe a note in the comment explaining that this traversale of subterms is unique in that it can affect annotations?